### PR TITLE
Add `deletionDate` to relationship's deletion info

### PR DESCRIPTION
**Pull request by sebbi08**
# Readiness checklist

-   [x] I added/updated tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

### CODE REVIEW

File: packages/runtime/test/transport/identityDeletionPeerProcessing.test.ts [hunk 1]
```diff
@@ -14,15 +14,15 @@ beforeAll(async () => {
     relationshipId = (await establishRelationship(services1.transport, services2.transport)).id;
 }, 30000);
 
+afterAll(async () => {
+    return await serviceProvider.stop();
```
Comment (jkoenig134) [inline packages/runtime/test/transport/identityDeletionPeerProcessing.test.ts:18]: Why did you move this? We always do afterAll after beforeAll, I don't see a reason to do it differently here.

File: packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts [hunk 2]
```diff
@@ -20,6 +23,10 @@ export class PeerDeletionInfo extends Serializable implements IPeerDeletionInfo
     })
     public deletionStatus: PeerDeletionStatus;
 
+    @serialize()
+    @validate({ nullable: true })
+    public deletionDate?: CoreDate;
```
Comment (jkoenig134) [inline packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts:28]: This should not be nullable and instead default to now in 14 Days when the contact is to be deleted or now if the contact is already deleted.

**Review by jkoenig134 — CHANGES_REQUESTED** [review]
I am missing one other external event processor regarding the identity deletion to be changed.

File: packages/runtime/test/transport/identityDeletionPeerProcessing.test.ts (hunk 1 above)
Comment (sebbi08) [reply packages/runtime/test/transport/identityDeletionPeerProcessing.test.ts:18]: Because before it was like

beforeAll
beforeEach
afterAll
afterEach

now it is 

beforeAll
afterAll
beforeEach
afterEach

we also have that in other tests.

File: packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts (hunk 2 above)
Comment (jkoenig134) [reply packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts]: > and instead default to now in 14 Days when the contact is to be deleted or now if the contact is already deleted

I think you forgot this whole part. Needs to be in preFrom.

Comment (sebbi08) [reply packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts]: I don't think we need the default, as we always get it from the corresponding External Event (PeerToBeDeleted and PeerDeleted). Or is there any other place where this default would be actually used ?

Comment (jkoenig134) [reply packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts]: Backwards compatibility.

File: packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts [hunk 3]
```diff
@@ -20,6 +23,17 @@ export class PeerDeletionInfo extends Serializable implements IPeerDeletionInfo
     })
     public deletionStatus: PeerDeletionStatus;
 
+    @serialize()
+    @validate()
+    public deletionDate: CoreDate;
+
+    public static override preFrom(value: any): any {
+        if (!value.deletionDate) {
+            value.deletionDate = value.deletionStatus === PeerDeletionStatus.ToBeDeleted ? CoreDate.local().add({ days: 14 }) : CoreDate.local();
```
Comment (jkoenig134) [inline packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts:32]: two things

1. switch over `value.deletionStatus as PeerDeletionStatus` to make sure the switch is exhaustive if we add another status
2. in each block do `CoreDate#toString()` to be closer to having the thing in the DB.

File: packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts [hunk 4]
```diff
@@ -35,7 +35,14 @@ describe("PeerDeletedExternalEventProcessor", function () {
...
+        await eventProcessor.execute({
+            id: "anId",
+            createdAt: "aDate",
+            index: 1,
+            syncErrorCount: 0,
+            type: "PeerDeleted",
+            payload: { relationshipId: relationshipId.toString(), deletionDate: CoreDate.local().add({ days: 24 }) }
+        });
         const relationship = await recipient.relationships.getRelationship(relationshipId);
         expect(relationship!.peerDeletionInfo!.deletionStatus).toBe(PeerDeletionStatus.Deleted);
```
Comment (jkoenig134) [inline packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts:47]: please add a test if the date was correctly saved

File: packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts [hunk 5]
```diff
@@ -41,7 +41,7 @@ describe("PeerToBeDeletedExternalEventProcessor", function () {
             index: 1,
             syncErrorCount: 0,
             type: "PeerToBeDeleted",
-            payload: { relationshipId: relationshipId.toString() }
+            payload: { relationshipId: relationshipId.toString(), deletionDate: CoreDate.local().add({ days: 14 }) }
         });
         const relationship = await recipient.relationships.getRelationship(relationshipId);
         expect(relationship!.peerDeletionInfo!.deletionStatus).toBe(PeerDeletionStatus.ToBeDeleted);
```
Comment (jkoenig134) [inline packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts:47]: please add a test if the date was correctly saved

**Review by jkoenig134 — CHANGES_REQUESTED** [review]

File: packages/transport/test/modules/relationships/PeerDeletionInfo.test.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        const peerDeletionInfoToBeDeleted = PeerDeletionInfo.fromAny({
+            deletionStatus: PeerDeletionStatus.ToBeDeleted
+        });
+        const peerDeletionInfoDeleted = PeerDeletionInfo.fromAny({
+            deletionStatus: PeerDeletionStatus.Deleted
+        });
+
+        expect(peerDeletionInfoToBeDeleted.deletionDate.isAfter(CoreDate.local())).toBeTruthy();
+        expect(peerDeletionInfoDeleted.deletionDate.isSameOrBefore(CoreDate.local())).toBeTruthy();
+    });
```
Comment (jkoenig134) [inline packages/transport/test/modules/relationships/PeerDeletionInfo.test.ts:15]: these are two tests.

File: packages/transport/test/modules/relationships/PeerDeletionInfo.test.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { CoreDate } from "@nmshd/core-types";
+import { PeerDeletionInfo, PeerDeletionStatus } from "../../../src";
+
+describe("PeerDeletionInfo", () => {
+    test("PeerDeletionInfo default values", () => {
```
Comment (jkoenig134) [inline packages/transport/test/modules/relationships/PeerDeletionInfo.test.ts:5]: this doesn't match our test naming and does not explain what should be tested here. I you would have written a proper test name you would have noticed that this test should be splitted in two tests.

File: packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts [hunk 8]
```diff
@@ -20,6 +23,23 @@ export class PeerDeletionInfo extends Serializable implements IPeerDeletionInfo
...
+
+    public static override preFrom(value: any): any {
+        switch (value.deletionStatus as PeerDeletionStatus) {
+            case PeerDeletionStatus.ToBeDeleted:
+                value.deletionDate = CoreDate.local().add({ days: 14 }).toString();
+                break;
+            case PeerDeletionStatus.Deleted:
+                value.deletionDate = CoreDate.local().toString();
+                break;
+        }
```
Comment (jkoenig134) [inline packages/transport/src/modules/relationships/local/PeerDeletionInfo.ts:38]: you introduced a bug by just deleting the if

File: packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts [hunk 9]
```diff
@@ -35,8 +35,16 @@ describe("PeerDeletedExternalEventProcessor", function () {
...
+            id: "anId",
+            createdAt: "aDate",
+            index: 1,
+            syncErrorCount: 0,
+            type: "PeerDeleted",
+            payload: { relationshipId: relationshipId.toString(), deletionDate: CoreDate.local() }
+        });
         const relationship = await recipient.relationships.getRelationship(relationshipId);
         expect(relationship!.peerDeletionInfo!.deletionStatus).toBe(PeerDeletionStatus.Deleted);
+        expect(relationship!.peerDeletionInfo!.deletionDate.isSameOrBefore(CoreDate.local())).toBeTruthy();
```
Comment (jkoenig134) [inline packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts:48]: why do you test it like this? you can put the date in a variable and check the exact date. Same for other test.

**Comment by sebbi08** [comment]
For this one yea in the other the date is create in the preFrom

**Comment by jkoenig134** [comment]
> For this one yea in the other the date is create in the preFrom

???

File: packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts (hunk 9 above)
Comment (sebbi08) [reply packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts]: For this one yea in the other the date is create in the preFrom

**Comment by sebbi08** [comment]
> > For this one yea in the other the date is create in the preFrom
> 
> ???

wow the reply function from the teams bot is not really working. Put the comment to the right thread

File: packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts (hunk 9 above)
Comment (jkoenig134) [reply packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts]: i meant the other processor. ANd you have already changed this.

File: packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts [hunk 10]
```diff
@@ -35,15 +35,20 @@ describe("PeerToBeDeletedExternalEventProcessor", function () {
...
             id: "anId",
             createdAt: "aDate",
             index: 1,
             syncErrorCount: 0,
             type: "PeerToBeDeleted",
-            payload: { relationshipId: relationshipId.toString() }
+            payload: {
+                relationshipId: relationshipId.toString(),
+                deletionDate: deletionDate
+            }
```
[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts:48]: you made the other one inline. Please decide for one...

[the remark this answers is not in the document]
Comment (jkoenig134) [reply packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts:48]: I am a big fan of inline in this case and it shouldn't be a problem with very small monitors.

**Review by jkoenig134 — DISMISSED** [review]

File: packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts [hunk 11]
```diff
@@ -35,8 +35,17 @@ describe("PeerDeletedExternalEventProcessor", function () {
...
         const eventProcessor = new PeerDeletedExternalEventProcessor(recipient.identityDeletionProcess.eventBus, recipient);
-        await eventProcessor.execute({ id: "anId", createdAt: "aDate", index: 1, syncErrorCount: 0, type: "PeerDeleted", payload: { relationshipId: relationshipId.toString() } });
+        const deletionDate = CoreDate.local();
+        await eventProcessor.execute({
+            id: "anId",
+            createdAt: "aDate",
+            index: 1,
+            syncErrorCount: 0,
+            type: "PeerDeleted",
+            payload: { relationshipId: relationshipId.toString(), deletionDate: deletionDate }
```
Comment (jkoenig134) [inline packages/transport/test/modules/sync/externalEventProcessors/PeerDeletedExternalEventProcessor.test.ts:45]: sorry, I didn't saw this. 

```suggestion
            payload: { relationshipId: relationshipId.toString(), deletionDate }
```

we always use the shorthand if possible.

File: packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (sebbi08) [reply packages/transport/test/modules/sync/externalEventProcessors/PeerToBeDeletedExternalEventProcessor.test.ts]: yea now i remebered why i did it because before the CoreDate creation was in there and that was way longer.

**Review by jkoenig134 — APPROVED** [review]
